### ISSUE TITLE
Better documentation on running pyodide in a browser

### ISSUE THREAD

**Issue by gregelin** [iodide-project/pyodide#586]
I tried to running Pyodide and Iodide without a server -- but it is unclear if it is possible or how to do either. There is a vague reference in documentation of how to including `pyodide.js` with a script tag, but it would be much better to have a working examples.

I would like to use Python in an environment where I cannot install ANY executable on the workstation -- no Docker, no Python, no Node, no Web server.

The Pyodide implementation and Iodide both seem very close to satisfying this need -- but it would be very helpful to have a full example -- or better an zip file with running code and everything local -- of running entirely locally in the browser.

**Comment by phorward** [comment]
1. Download a Pyodide [release tarball](https://github.com/iodide-project/pyodide/releases), like [pyodide-0.14.3.tar.bz2](https://github.com/iodide-project/pyodide/releases/download/0.14.3/pyodide-build-0.14.3.tar.bz2)
2. Extract the downloaded archive (e.g. `mkdir pyodide && tar -xjvf pyodide-build-0.14.3.tar.bz2  -C pyodide/`
3. Goto `pyodide/` folder with the extracted files and create a `index.html` with the following content
```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Pyodide Minimal Example</title>

    <script>
		window.languagePluginUrl = "./";
    </script>
    <script src="./pyodide.js"></script>
</head>
<body>
<script>
	languagePluginLoader.then(() => {
		window.pyodide.runPython(`
print("Hello World")
`);
	});

</script>

</body>
</html>
```
4. Start a HTTP server (e.g. `python3 -m http.server`)
5. Visit http://localhost:8000/ and look into the JavaScript console: `Hello World`

> I would like to use Python in an environment where I cannot install ANY executable on the workstation -- no Docker, no Python, no Node, no Web server.

That's exactly how we got it to run. Our customer uses thin clients with Chrome and Windows only, the Pyodide-based Python-applications is loaded from a website hosted in the cloud. But you need some kind of server, which serves the application you want to run to the client's browser. This can be on the internet or on an intranet.

Step 4 can also be the upload to a HTTP/HTTPS webserver, or cloud service. You may require more configuration of the server's MIME types because wasm-files must be delivered as `application/wasm`, which unfortunately isn't standard, yet.

**Comment by zeluspudding** [comment]
Here are two examples of Pyodide live on codepen: [Hello World](https://codepen.io/aagostini/pen/gObgpvz), [Matplotlib Plot](https://codepen.io/aagostini/pen/LYExVJL?editors=1000)... courtesy of @ImportanceOfBeingErnest on [StackOverflow](https://stackoverflow.com/a/56670502/3491991)

**Comment by gregelin** [comment]
Thanks @phorward. I did manage to get that far. Unfortunately, the environment I am targeting needs files opened via `file://` rather than `http://` (e.g., no web server). 

Thanks also @zeluspudding

**Comment by bil-ash** [comment]
> 1. Download a Pyodide [release tarball](https://github.com/iodide-project/pyodide/releases), like [pyodide-0.14.3.tar.bz2](https://github.com/iodide-project/pyodide/releases/download/0.14.3/pyodide-build-0.14.3.tar.bz2)
> 2. Extract the downloaded archive (e.g. `mkdir pyodide && tar -xjvf pyodide-build-0.14.3.tar.bz2  -C pyodide/`
> 3. Goto `pyodide/` folder with the extracted files and create a `index.html` with the following content
> 
> ```
> <!DOCTYPE html>
> <html>
> <head>
>     <meta charset="utf-8">
>     <title>Pyodide Minimal Example</title>
> 
>     <script>
> 		window.languagePluginUrl = "./";
>     </script>
>     <script src="./pyodide.js"></script>
> </head>
> <body>
> <script>
> 	languagePluginLoader.then(() => {
> 		window.pyodide.runPython(`
> print("Hello World")
> `);
> 	});
> 
> </script>
> 
> </body>
> </html>
> ```
> 
> 1. Start a HTTP server (e.g. `python3 -m http.server`)
> 2. Visit http://localhost:8000/ and look into the JavaScript console: `Hello World`
> 
> > I would like to use Python in an environment where I cannot install ANY executable on the workstation -- no Docker, no Python, no Node, no Web server.
> 
> That's exactly how we got it to run. Our customer uses thin clients with Chrome and Windows only, the Pyodide-based Python-application is loaded from a website hosted in the cloud. But you need some kind of server, which serves the application you want to run on the client's browser. This can be on the internet or on an intranet service.
> 
> Step 4 can also be the upload to a HTTP/HTTPS webserver, or cloud service. You may require more configuration of the server's MIME types because wasm-files must be delivered as `application/wasm`, which unfortunately isn't standard, yet. [Here](https://gist.github.com/phorward/02cc6f73eff06c8e8397ebb34f98e7dc) I posted configuration examples for Apache Web Server and Google App Engine as a hint.

Quite lucid,this should actually be added to the docs

**Comment by bil-ash** [comment]
https://github.com/iodide-project/pyodide/issues/586#issuecomment-566084111
Quite lucid, this should be added to the docs

**Comment by basvandertol** [comment]
> I would like to use Python in an environment where I cannot install ANY executable on the workstation -- no Docker, no Python, no Node, no Web server.

Does that exclude browser extensions as well? A browser extension could mimic a server by intercepting web requests, and returning local files instead. But that is of no help if you can not install an extension.

**Comment by gregelin** [comment]
Yes, the environment would also exclude installing browser extensions.

Greg Elin
CEO, GovReady PBC
p: 917-304-3488
e: gregelin@govready.com

**Comment by mihalski** [comment]
I am in the same situation as @gregelin.

Is there a solution that we can utilise or is our scenario outside the scope of this project?

**Comment by basvandertol** [comment]
@mihalski and @gregelin
I have made some progress in the direction of a solution.
The problem is that for security reasons a script running in a browser can not open a local url (file://)
without explicit permission of the user for each file.
A solution is to encode the binary files in base64 and embed them in a .html file
I have managed to do this for the pyodide.asm.wasm and could subsequently load the file as ReadableStream from a local url and compile it with a call to WebAssembly.compileStreaming().

I am planning to work out a full solution. This is just a hobby project. I am not a member of the pyodide development team. I can not give you a timeline, but over the weekend I'll put what I have got so far in a pull request (as work in progress)

**Comment by mihalski** [comment]
@basvandertol Thank you for the update! We all appreciate the work you're putting in :)

**Comment by basvandertol** [comment]
It works! Scripts to create a local pyodide version are in #606

**Comment by basvandertol** [comment]
See the [full example](https://github.com/basvandertol/pyodide/releases/download/localpyodide-v0.1/localpyodide.zip) (218MB zip file)

**Comment by phorward** [comment]
Thanks, this is really cool. I'll take a closer look into it soon. The feature is absolutely useful, not even for local environments.

**Comment by MaBuStarr** [comment]
https://pyodide.org/en/stable/console.html check this out.
